tests/Progress/ProgressUpdateRepositoryTest.php

```diff
@@ -0,0 +1,183 @@
+<?php
+
+namespace BristolSU\Support\Tests\Progress;
+
+use BristolSU\Support\Activity\Activity;
+use BristolSU\Support\ActivityInstance\ActivityInstance;
+use BristolSU\Support\ModuleInstance\ModuleInstance;
+use BristolSU\Support\Progress\ModuleInstanceProgress;
+use BristolSU\Support\Progress\Progress;
+use BristolSU\Support\Progress\ProgressHashes;
+use BristolSU\Support\Progress\ProgressUpdateRepository;
+use BristolSU\Support\Progress\Snapshot;
+use BristolSU\Support\Tests\TestCase;
+use Carbon\Carbon;
+
+class ProgressUpdateRepositoryTest extends TestCase
+{
+    protected ProgressUpdateRepository $repository;
+    protected $activity;
+
+    /**
+     * @var ModuleInstance[] $modules
+     */
+    protected $modules;
+    protected $activityInstance;
+    protected $hashesTableName;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->repository = new ProgressUpdateRepository();
+
+        $this->activity = factory(Activity::class)->create();
+        $this->modules = factory(ModuleInstance::class, 5)->create(['activity_id' => $this->activity->id]);
+        $this->activityInstance = factory(ActivityInstance::class)->create(['activity_id' => $this->activity->id]);
+
+        $this->hashesTableName = (new ProgressHashes)->getTable();
+    }
+
+    /** @test */
+    public function saving_progress_saves_a_new_hash_correctly()
+    {
+        // Create a new Snapshot Instance and trigger Update to Activity Instance:
+        (new Snapshot($this->repository))->ofUpdateToActivityInstance($this->activityInstance, 'caller');
```

Comment:
I think this is slightly overcomplicated. As with the Snapshot test, we want to make sure that ONLY a single class is being tested. With this, a change in the Snapshot class could break this test.

You don't need to make reference to Snapshot anywhere in this class. Just get the repository, and call `$this->repository->saveChanges()` directly then assert as you already are!